_overviews/scala3-book/domain-modeling-fp.md

```diff
@@ -0,0 +1,408 @@
+
+
+
+
+## What about the operations?
+
+This leads to an interesting question: Because FP separates the data from the operations on that data, how do you implement those operations in Scala/FP?
+
+There are several different ways to handle behaviors:
+
+- Put your methods in companion objects
+- Use a modular programming style
+- Use a “functional objects” approach
+
+These solutions are shown in the next three sections.
+<!-- TODO: is there ONE preferred solution? -->
+
+
+>An important point about all of the following solutions is that the methods and functions shown are *pure functions*. They don’t mutate any data; they receive an object, and apply transformation functions to that data to return new data.
+<!-- TODO: link to "pure function" definition -->
+
+
+
+
+## (1) Handling behavior with a companion object
+
+A first approach is to put the behaviors — the methods — in a companion object. With this approach, you put the data model in a `case` class, and then put the methods in a companion object of the `case` class.
+
+The following code shows how to add and remove toppings, and update the crust size and crust type for an FP-style `Pizza`:
+
+```scala
+case class Pizza(
+  crustSize: CrustSize,
+  crustType: CrustType,
+  toppings: Seq[Topping]
+)
+
+object Pizza:
+  def addTopping(p: Pizza, t: Topping): Pizza =
+    val newToppings = p.toppings :+ t
+    p.copy(toppings = newToppings)
+
+  def removeTopping(p: Pizza, t: Topping): Pizza =
+    val newToppings = ListUtils.dropFirstMatch(p.toppings, t)
+    p.copy(toppings = newToppings)
+
+  def removeAllToppings(p: Pizza): Pizza =
+    val newToppings = Seq[Topping]()
+    p.copy(toppings = newToppings)
+
+  def updateCrustSize(p: Pizza, cs: CrustSize): Pizza =
+    p.copy(crustSize = cs)
+
+  def updateCrustType(p: Pizza, ct: CrustType): Pizza =
+    p.copy(crustType = ct)
+```
+
+With this approach you can create an initial `Pizza` like this:
+
+```scala
+val pizza1 = Pizza(Small, Thin, Seq(Cheese, Onions))
+```
+
+Then as a customer edits their order, you create new `Pizza` instances by making changes to an initial object, while assigning the result to a new variable:
+
+```scala
+val pizza2 = Pizza.addTopping(pizza1, Pepperoni)
+val pizza3 = Pizza.updateCrustSize(pizza2, Large)
+```
+
+If you prefer to get rid of all those `Pizza` references, you can eliminate them by importing the methods from the `Pizza` companion object:
+
+````
+import Pizza._
+
+val pizza1 = Pizza(Small, Thin, Seq(Cheese, Onions))
+val pizza2 = addTopping(pizza1, Pepperoni)
+val pizza3 = updateCrustSize(pizza2, Large)
+````
+
+
+
+## (2) Handling behavior with a modular approach
```

Comment:
This title is a bit too vague imho.
You are doing the same as above but following the contract specified in a trait.

Reply:
I think Jonathan’s changes address this, so I’m resolving this comment.